core_impl/src/main/java/com/google/instrumentation/trace/SpanImpl.java

```diff
@@ -303,20 +330,45 @@ public void end(EndSpanOptions options) {
     void onEnd(SpanImpl span);
   }
 
+  private static final class AttributesWithCapacity extends LinkedHashMap<String, AttributeValue> {
+    private int capacity;
+    private int totalRecordedAttributes = 0;
+    private static final long serialVersionUID = 42L;
+
+    private AttributesWithCapacity(int capacity) {
+      super(capacity);
+      this.capacity = capacity;
+    }
+
+    private void addAttributes(Map<String, AttributeValue> attributes) {
+      totalRecordedAttributes += attributes.size();
+      putAll(attributes);
+    }
+
+    private int getNumberOfDroppedAttributes() {
+      return totalRecordedAttributes - size();
+    }
+
+    @Override
+    protected boolean removeEldestEntry(Map.Entry<String, AttributeValue> eldest) {
```

Comment:
It would also help to have a comment explaining this class, i.e., the reason it extends LinkedHashMap and overrides removeEldestEntry.